### ISSUE TITLE
`declaration-block-no-redundant-longhand-properties` is not forwards compatible

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#7630]
Forwards compatibility means that future CSS can be used in a (slightly) older version of Stylelint.

I don't expect this to be true for large syntax changes.
But I expect Stylelint to not break my code each time a new constituent is added to a shorthand.

A concrete example is the introduction of `text-decoration-thickness` in `text-decoration`.
The syntax was updated:
- from `<'text-decoration-line'> || <'text-decoration-style'> || <'text-decoration-color'>`
- to `<'text-decoration-line'> || <'text-decoration-thickness'> || <'text-decoration-style'> || <'text-decoration-color'>`

See :

https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdugLjmAHjgLSyQBOSOAlhJsTgBbVgDWmMAzp-CgEwAHAgG4sufEVIwKVWvU44AngBsYvKBDwwoo7FslkIlGnWLLqHXgFdMscuY67xhEoeNzikZUd7ltugF8QABoQADNqVQA5JABbOERCOIFVADowbhDwOgiAcwQQDEwUFHQQFxhbTjLeAG0sEpKyhRUYBxJITDziBSRbJHIoMoaUAF0sINDOvIAxI1iqAoArTjos2AFOAqKmkBbVdprSkGUqLhwy4JHmpQOLDpzqXJ6cPqgBoYRj0-wFYcwggEgA

```css
a {
	text-decoration-thickness: 2px;
	text-decoration-style: dotted;
	text-decoration-line: underline;
	text-decoration-color: red;
}
```

Is changed to 

```css
a {
	text-decoration-thickness: 2px;
	text-decoration: underline dotted red;
}
```

Which erases `text-decoration-thickness`.

Updates can be made to teach Stylelint about the updated syntax but this implies a significant maintenance burden.

This is different from other kinds of unknown syntax or CSS aspects as many CSS authors will not have memorized exactly which longhands are constituents of which shorthands and might not even notice the destructive changes Stylelint makes in these cases.

------

The issue here is not `text-decoration` itself, but that changing longhands to shorthands is a very impactful source code change.

Longhands and shorthands are simply not the same in CSS.

This issue intended for discussion around this topic and to make sure that this aspect of this rule is in line with the overal goals of Stylelint core rules.

**Comment by romainmenke** [comment]
Extra disclosure :

I don't use this rule and actually actively push team members to use longhands over shorthands. I want this rule to work well for those who want to use it, but I am not attached to or invested in it.

**Comment by Mouvedia** [comment]
Concerning #7232, I have introduced an option with the fix so that the impact can be mitigated.
Concerning large scale impacts, I have advised to update `stylelint/stylelint-config-standard` with `"ignoreLonghands": ["text-decoration-thickness"]` as a fail-safe in [this comment](https://github.com/stylelint/stylelint/pull/7611#issuecomment-2061820286).

> changing longhands to shorthands is a very impactful source code change

I think I did my due diligence in that issue to explain the impact in minute details and introducing a secondary option with the fix.

> This issue intended for discussion around this topic and to make sure that this aspect of this rule is in line with the overal goals of Stylelint core rules.

Should this be moved to _discussions_?
If so can it be made public?
I personally don't have access to it.

> For this point, @Mouvedia may have any thoughts.

I think the policy should be to introduce changes if the support reaches a certain threshold.
In the case of `text-decoration-thickness` it's above 95% according to caniuse.
In that regard I think the discussion we had on #301 is relevant.
i.e. IMHO it's much more impactful for `stylelint/stylelint-config-standard` and `stylelint/stylelint-config-recommended`

From a maintainer stand point, it's a matter of being careful about that type of changes.
e.g. not accepting a fix without an accompanied ignore option

**Comment by ybiquitous** [comment]
> Should this be moved to _discussions_?

This has already the `needs discussion` label, right?

**Comment by ybiquitous** [comment]
> I was talking about [docs.github.com/en/discussions](https://docs.github.com/en/discussions)

We don't run GitHub Discussions since we've managed discussions by the label so far. This will not change in the future.

**Comment by ybiquitous** [comment]
The problem here is not only specific properties (e.g. `text-decoration-thickness`), but also essentially difficult to handle transformation between longhands and shorthands.

If we can't easily resolve the problem, we should remove the rule from our shared config and even leave a warning in the rule doc, at least until the rule would become enough mature.

**Comment by ybiquitous** [comment]
When seeing the rule's implementation, custom logic for some properties have increased. Such a logic will be more, as new properties are added to CSS.

https://github.com/stylelint/stylelint/blob/9afaa08c196626fac36e9f6b8b8db9dc164cc11d/lib/rules/declaration-block-no-redundant-longhand-properties/index.mjs#L28

**Comment by romainmenke** [comment]
> `"ignoreLonghands": ["text-decoration-thickness"]`

This is indeed very useful as an escape hatch when user notice that their CSS is altered in incorrect ways.

But it does expect users to notice. I fear that most will assume that Stylelint knows best and will get frustrated when they eventually find that Stylelint is breaking their code.

We could have automation/signals to help us update this rule more quickly ([`@webref/css`](https://www.npmjs.com/package/@webref/css?) has data for longhands, shorthands, logical groups, ...)

---------------

Ultimately this is also an upstream issue :)
Adding constituents to a shorthand or making a longhand into a shorthand is a breaking change in CSS itself. So it also breaks downstream tools (e.g. Stylelint)

**Comment by ybiquitous** [comment]
> But it does expect users to notice. I fear that most will assume that Stylelint knows best and will get frustrated when they eventually find that Stylelint is breaking their code.

Ah, this seems likely to me... We may need to even suspend the rule's autofix.

**Comment by ybiquitous** [comment]
> At the minimum add `"ignoreLonghands": ["text-decoration-thickness"]`.

I think this will not resolve the fundamental problem with this rule. First of all, we should discuss if this rule should be present the shared config.

**Comment by romainmenke** [comment]
Aside:

Wanted to say that none of this is feedback on the hard work that so many people have been putting into this rule :)

It is a really complicated subject that is non-trivial to get right and maintain.

This is mostly an illustration of how complexity cascades. Shorthands are significant complexity in CSS itself and any tooling that touches on shorthands will also encounter complexity.

**Comment by romainmenke** [comment]
See : https://github.com/stylelint/stylelint/issues/7917#issuecomment-2282808664

Aside from the autofix sometimes producing CSS where the shorthand is placed before unknown constituent properties there is also a second issue.

Longhand properties are (in)validated separately while a longhand is valid only when the entirety is valid.

 ```css
a {
	text-decoration-thickness: 2px; /* only valid in modern browsers */
	text-decoration-style: dotted; /* valid in most browsers */
	text-decoration-line: underline; /* valid in most browsers */
	text-decoration-color: red; /* valid in most browsers */
}
```

```css
a {
	text-decoration: underline dotted red 2px; /* only valid in modern browsers */
}
```

Mutating CSS to use more shorthands also alters the support.
This might be very disruptive.

**Comment by romainmenke** [comment]
see: https://github.com/stylelint/stylelint/issues/7917

---------

Honestly I feel that `declaration-block-no-redundant-longhand-properties` is very complicated to maintain and support.

We often have issues here and discuss the same things with regard to these rules.

Given the time taken up by `declaration-block-no-redundant-longhand-properties` and the limited resources of Stylelint maintainers I do feel we need to take some action here.

I think the core issue is that these are very different rules, with very different constraints and goals.

`declaration-block-no-shorthand-property-overrides`:
- warns when CSS declarations can never have an effect
- is factual
- is not affected by browser support
- is low risk

`declaration-block-no-redundant-longhand-properties`:
- warns when CSS could be written as a one liner
- is opinionated/stylistic
- is heavily dependent on browser support
- makes me very worried

-------

Can we embrace these differences?

Each rule has its own data set.
Each rule can be updated without affecting the other.

We always try to be up to date with recent CSS additions in `declaration-block-no-shorthand-property-overrides`.

We should respect browser support concerns for `declaration-block-no-redundant-longhand-properties`.

-------

At this time I see two possible directions:
- change the scope and purpose of `declaration-block-no-redundant-longhand-properties`
- drop `declaration-block-no-redundant-longhand-properties`

I think that `declaration-block-no-shorthand-property-overrides` is fine and can remain as it exists today?

-------

### Alternative proposal for `declaration-block-no-redundant-longhand-properties`?

Currently this rule works kinda like: "everything that can be written as a shorthand must be written as a shorthand".

But it doesn't consider browser support, readability, ...

Arguably most things aren't better as shorthands at all:
- `font: italic normal bold normal .8em 1.2 Arial;`
- `border-width: 1px 2px 3px 4px;`

I think that some very specific patterns are really handy:
- `padding: 10px 20px`
- `border-top: 1px solid green`

But I think the bar should be pretty high for automatically generated CSS.
There should be broad consensus that the generated CSS is good.

Should we instead focus on these specific patterns that are widely used and agreed upon to be good and readable? Instead of trying to convert everything possible to shorthand, to only have a very specific set?

This makes the rule even more opinionated but makes the pace of change much slower.

_These transform should (obviously) never emit the shorthands after other constituents. Doing so anyway would always be considered a bug._

### Drop `declaration-block-no-redundant-longhand-properties` as other stylistic rules have been dropped in the past?

This rule doesn't help authors write correct or modern CSS.
It is very opinionated and arguably even stylistic.

By its very nature it can not fix broken or mistaken CSS. However it often produces broken CSS. This isn't good design for a core Stylelint rule.

I think we should also consider if this rule still aligns with the goals and priorities of Stylelint.

**Comment by Mouvedia** [comment]
I am of the opinion that your stance is a bit extreme.
What I would advice is:

1. fix #7609
2. add a new enhancement issue for partial autofixing
see https://github.com/stylelint/stylelint/issues/7232#issuecomment-2061880291
3. add a new bug issue to cover the reordering of issues while fixing
see https://github.com/stylelint/stylelint/issues/7917#issuecomment-2282812541
4. an a new issue on `stylelint-standard-config` that discuss whether we should add [`{ "disableFix": true }`](https://stylelint.io/user-guide/configure/#disablefix) to `declaration-block-no-redundant-longhand-properties` until 1., 2. and 3. are resolved

tl;dr: I think the autofix of this rule has minor defects that need to fixed non-urgently

**Comment by Mouvedia** [comment]
> Maybe each rule has its own data set?

As discussed while fixing #7610, if we have an issue that requires to not have a single source of truth, this refactor would be warranted.
At a minimum, we should have 2 map exported sharing the same code, instead of 1 currently.
i.e. do the refactor while fixing a concrete bug

**Comment by ybiquitous** [comment]
I also think dropping `declaration-block-no-redundant-longhand-properties` is too much because some may still consider the rule helpful. But I also agree that this rule sometimes requires us to be careful when using it.

So, as @Mouvedia advised on <https://github.com/stylelint/stylelint/issues/7630#issuecomment-2282856579>, it seems better to me to adjust options for `declaration-block-no-redundant-longhand-properties` in `stylelint-standard-config`. E.g.

```diff
-		'declaration-block-no-redundant-longhand-properties': true,
+		'declaration-block-no-redundant-longhand-properties': [true, {
+			ignoreLonghands: [/* still immature properties in most browsers */],
+			ignoreShorthands: [/* still immature properties in most browsers */],
+			disableFix: true, // prevents incorrect autofix
+		}],
```

-- https://github.com/stylelint/stylelint-config-standard/blob/d319b67e7b8a546bca84c6a909764171df587d7e/index.js#L55

Ref https://stylelint.io/user-guide/rules/declaration-block-no-redundant-longhand-properties

What do you think? @romainmenke

**Comment by romainmenke** [comment]
> I am of the opinion that your stance is a bit extreme.

> I also think dropping declaration-block-no-redundant-longhand-properties is too much because some may still consider the rule helpful. But I also agree that this rule sometimes requires us to be careful when using it.

That is fair :)

I assign very high value to dev tools that work without friction and equally consider dev tools that produce broken source code as highly problematic. In reality however this is more nuanced and most processes involve a human that can spot weird outcomes. We have options to disable rules, ignore specific errors with configuration commands, ...

------

How do such options on `stylelint-config-standard` work with overrides set by users?
If a user sets their own `ignoreLonghands`, `ignoreShorthands` values, do these merge or override with the options from `stylelint-config-standard`?

It seems like quite a hassle if they override and users need to alter these each time we do. But I don't know how they work, so maybe a non-issue?

--------

I was looking through the rules and I was wondering if we could do something like this:
https://github.com/stylelint/stylelint/pull/7924

Where we store "is baseline widely available" into the mapping of properties.
These currently are `Map<string, Set<String>>` but they could also be `Map<string, Map<string, boolean>>`, where the `boolean` indicates the baseline status.

This allows us to have a single source of truth.

https://github.com/stylelint/stylelint/blob/4a71e6c7304c567d4dbf3418cc7a78fb2ba67082/lib/reference/properties.mjs#L609

Is this maybe also an option?

**Comment by Mouvedia** [comment]
What you are describing looks like feature flags to me.
i.e. most of the time it gives the user the liberty of passing a config to override the defaults
That sounds like a huge undertaking so let's forget about that aspect for now.
Let's K.I.S.S.

The idea of being able to start to work on a feature without a worry because it's behind a flag is enticing to me.
I won't discuss implementation details in the issue though.

**Comment by ybiquitous** [comment]
@romainmenke If users want to customize a rule's settings defined in a config like `stylelint-config-standard`, they need to override the settings; then cannot merge them.

For example, if we want to customize `at-rule-empty-line-before`, we need to write our `stylelint.config.mjs` with:

```js
export default {
	extends: 'stylelint-config-standard',
	rules: {
		'at-rule-empty-line-before': ['never', { ignore: ['first-nested'] }],
	},
};
```

```sh-session
$ npx stylelint --print-config stylelint.config.mjs | jq '.rules["at-rule-empty-line-before"]'
[
  "always",
  {
    "except": [
      "blockless-after-same-name-blockless",
      "first-nested"
    ],
    "ignore": [
      "after-comment"
    ]
  }
]
```

If there is no customization, then:

```js
export default {
	extends: 'stylelint-config-standard',
};
```

```sh-session
$ npx stylelint --print-config stylelint.config.mjs | jq '.rules["at-rule-empty-line-before"]'
[
  "always",
  {
    "except": [
      "blockless-after-same-name-blockless",
      "first-nested"
    ],
    "ignore": [
      "after-comment"
    ]
  }
]
```

-- https://github.com/stylelint/stylelint-config-standard/blob/d319b67e7b8a546bca84c6a909764171df587d7e/index.js#L18-L24

**Comment by ybiquitous** [comment]
Of course, instead of adjusting `declaration-block-no-redundant-longhand-properties` with the secondary options (e.g., `ignoreLonghands`) in `stylelint-config-standard`, we can handle some special cases, as #7924 does for `transition-behavior`.

This way seems straightforward for us (maintainers), however, there is a downside that users cannot take advantage of `transition-behavior` linting, although this downside may be trivial.

In addition, we have already warned in the rule's README like this:

https://github.com/stylelint/stylelint/blob/1fba291f3b7e88970d54566b360841cd123fc565/lib/rules/declaration-block-no-redundant-longhand-properties/README.md?plain=1#L88-L89

Disabling autofix for the rule in `stylelint-config-standard` seems still reasonable to me:

```diff
-		'declaration-block-no-redundant-longhand-properties': true,
+		'declaration-block-no-redundant-longhand-properties': [true, {
+			disableFix: true, // prevents incorrect autofix
+		}],
```

**Comment by ybiquitous** [comment]
In summary:

- I agree with adding `transition-behavior` to `reference.longhandSubPropertiesOfShorthandProperties` for `declaration-block-no-shorthand-property-overrides` bug fixes.
- but I disagree with changing the current data structure of `longhandSubPropertiesOfShorthandProperties` because it's already exposed, and the structure change would become inconsistent with other reference data:
  https://github.com/stylelint/stylelint/blob/1fba291f3b7e88970d54566b360841cd123fc565/types/stylelint/index.d.ts#L859-L861
- so, we should handle special cases like `transition-behavior` only in `declaration-block-no-redundant-longhand-properties` if we could accept the implementation/maintenance cost.
- besides, it seems better to disable autofix `declaration-block-no-redundant-longhand-properties` in `stylelint-config-standard`.

**Comment by romainmenke** [comment]
> Disabling `autofix` for the rule in `stylelint-config-standard` seems still reasonable to me

I agree :)

This makes it less likely that an unwanted change is overlooked.

> but I disagree with changing the current data structure of `longhandSubPropertiesOfShorthandProperties` because it's already exposed

Yes, also not sure about that, but there are other ways to get similar results in the rule :)

----

I am not sure how to balance these:
- users might only target very modern browser versions and want all the shorthands that are available
- users want to target more diverse browser versions and don't want fixes that break their pages in some browser versions

I think that using `ignoreLonghands` and `ignoreShorthands` is the only way to support both use cases.
Even if it places a (slightly) higher burden on those where the impact is most negative.

I personally favor a system that is more cautious, but I am fine with only the change in the shared configs :)

**Comment by romainmenke** [comment]
Maybe this is a good approach for this rule:
- we always try to be up to date with our mapping of shorthands and longhands
- we use `ignoreLonghands` and  `ignoreShorthands` in the shared configs for those properties that aren't widely available yet
- when we receive bug reports for specific cases we add special handling in `declaration-block-no-redundant-longhand-properties`

**Comment by ybiquitous** [comment]
@romainmenke Thanks for summarizing. I'm in favor of the summary <https://github.com/stylelint/stylelint/issues/7630#issuecomment-2283431464>. 👍🏼

**Comment by Mouvedia** [comment]
Since we are about to reach a resolution, I would like to say that Id prefer if impactful changes to `stylelint-standard-config` were all contained in the same version.

So to be clear based on

https://chromestatus.com/metrics/css/popularity:
<sup>i.e. completely arbitrary rule of mine: above 55% themselves and their counterpart being under 40%</sup>

- `-webkit-appearance`
- `-webkit-text-size-adjust`
- `-webkit-font-smoothing`

our discussion in #7917:

- `transition-behavior`

whatwg/compat#28 and #7727:

- `-webkit-background-size`

I would send several PRs to add exceptions so that they can be merged in the same timespan.
Let's continue our discussion in a new issue on `stylelint-standard-config`'s side.

@romainmenke Can we close this is issue?

**Comment by romainmenke** [comment]
Yes, this issue can be closed, thank you 🙇

**Comment by ybiquitous** [comment]
Opened an issue:
https://github.com/stylelint/stylelint-config-standard/issues/332
